### ISSUE TITLE
`reportUnknownVariableType` sometimes not reported

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#10580]
**Describe the bug**

`reportUnknownVariableType` is sometimes not reported.

In the `foo` function below, `r` is revealed (in the mouse-over) as `dict[Unknown, Unknown]`, but the unknown type is not reported.

**Code or Screenshots**

```python
from typing import TypedDict


class TypedDictA(TypedDict):
    a: int


class TypedDictB(TypedDict):
    r: dict[str, TypedDictA]


def foo() -> TypedDictB:
    # no reportUnknownVariableType triggered anywhere here
    r = {}
    r["r_key"] = {
        "a": "a",
    }
    return {
        "r": r,
    }


def goo() -> TypedDictA:
    p = {}
    return p  # correctly triggers reportUnknownVariableType
```

pyright playground link:

https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjAFD1UA2AhgM6uEkXUwCCAFEQRlKNAJQAuWlBlRmEzCjoMW7TsO40AQoK6iYk6bJALSPANqsYIADTqRPXgF16tUsWBRgYMPzFQAWgA%2Be00YLSlZKABiKBQwKBBiHDwAVRQAa3iAdxQANWYQJGYAI0ZiIWJYIrQ0YiTSORQ4bIALeqr2pKMZECgAXigAbwBfHsTzACIQAH0M4jhJpwHh8ajJ5kmFDcmbcbGopJgAVxAUVaj1kC3Evaix%2BndPNB8-QJDKsN5I2QQV0fGR1O5z%2BMSgVFwSRojDg1SQtXqHCSKRg6SyYFyBSKpXKlVoQA

**Comment by erictraut** [comment]
This behavior is specific to empty containers (lists or dicts) that are created with a `[]` or `{}` expression. I originally designed it this way to accommodate patterns like this:

```python
def func1():
    val1 = []

    if a:
        val1 = [2, 3]

    reveal_type(val1) # list[int]
```

Pyright has some special handling for empty containers created with `[]`, `{}`. Their type is initially `list[Unknown]` and `dict[Unknown, Unknown]`, but if this is merged with another list or dict with a known type, the merge elides the "unknown" version from the resulting union. This is type safe, and it's a good convenience feature.

The original thinking was that the initial assignment of the empty container should be exempt from the `reportUnknownVariableType` check, but this can result in type checking holes as you've demonstrated. I wrote this in the early days of pyright (about five years ago) prior to working out all of the principles and rules for strict-mode type checking, and I had forgotten about this exemption until I started to dig into your bug report.

Perhaps the best thing to do is to simply remove this exemption. I'm a little worried that this change will result in a lot of noise (true false positives) for code bases that use pyright's strict type checking mode, but it's probably the right thing to do because the current logic allows for false negatives.

I'm going to think about this a bit more before proceeding. Let me know if you have any suggestions or preferences.

**Comment by beauxq** [comment]
On the question of whether to report the unknown on the initial assignment, I know it would be a change for many pyright strict users, but it might be a welcome change.

When I'm dealing with a pattern like this:
```python
a = []
a.append(1)
```
It has bothered me a little that the `unknown` report is on the `append` line and not on the assignment line, because the assignment is where I change something to deal with it (add a type annotation).
So reporting the problem on the assignment would not be bad in my view, because that is the line where something changes to address the diagnostic.

It surprises me that mypy even reports this problem without --strict
```python
r = {}  # mypy error: "need type annotation"
```
I don't think that error needs to be in the default configuration, but I do think it is valuable for strict mode.

**Comment by beauxq** [comment]
Overall, I think mypy handles issues around this well (except for the error on the assignment without strict mode).
```python
def foo() -> TypedDictB:
    r = {}
    # ^ mypy reports "need type annotation" here
    # I think it would be better to use a more strict rule for this,
    # but mypy reports it even without --strict

    reveal_type(r)  # dict[Any, Any]
    r["r_key"] = {
        "a": "a",
    }
    reveal_type(r)  # dict[str, dict[str, str]]

    # When there's no annotation, only the first __setitem__ affects the type.
    # The 2nd __setitem__ is tested against that type.
    r[2] = {"b": "b"}
    # ^ mypy reports "Invalid index type"
    # I think this is valuable.
    # It's likely a type error, and in the rare cases when it isn't, the user
    # can address it by adding an annotation to the first assignment.
    # If we instead change the type to dict[str | int, ...
    # then we have the danger of a false negative `if "2" in r:`

    return {
        "r": r,  # mypy gives an appropriate "incompatible type" error here
    }
```
